scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -1919,30 +2179,117 @@ void BPF_STRUCT_OPS(lavd_enqueue, struct task_struct *p, u64 enq_flags)
 	put_global_rq(p, taskc, cpuc, enq_flags);
 }
 
+static bool is_kernel_task(struct task_struct *p)
+{
+	return p->mm == NULL;
+}
+
 void BPF_STRUCT_OPS(lavd_dispatch, s32 cpu, struct task_struct *prev)
 {
+	struct bpf_cpumask *active, *ovrflw;
+	struct task_struct *p;
+
+	bpf_rcu_read_lock();
+
+	/*
+	 * Prepare cpumasks.
+	 */
+	active = active_cpumask;
+	ovrflw = ovrflw_cpumask;
+	if (!active || !ovrflw) {
+		scx_bpf_error("Failed to prepare cpumasks.");
+		goto unlock_out;
+	}
+
+	/*
+	 * If the CPU belonges to the active or overflow set, dispatch a task.
+	 */
+	if (bpf_cpumask_test_cpu(cpu, cast_mask(active)) ||
+	    bpf_cpumask_test_cpu(cpu, cast_mask(ovrflw))) {
+		scx_bpf_consume(LAVD_GLOBAL_DSQ);
+		goto unlock_out;
+	}
+
 	/*
-	 * Now, a CPU has no task to run, so it gets a task from the global run
-	 * queue for execution.
+	 * If this CPU is not either in active or overflow CPUs, it tries to
+	 * find and run a task pinned to run on this CPU.
 	 */
-	scx_bpf_consume(LAVD_GLOBAL_DSQ);
+	__COMPAT_DSQ_FOR_EACH(p, LAVD_GLOBAL_DSQ, 0) {
+		/*
+		 * Prioritize kernel tasks because most kernel tasks are pinned
+		 * to a paritular CPU and latency-critical (e.g., ksoftirqd,
+		 * kworker, etc).
+		 */
+		if (is_kernel_task(p)) {
+			scx_bpf_consume(LAVD_GLOBAL_DSQ);
```

Comment:
Do we want to consume the task here? I think you'd have to do something like:

```
if (__COMPAT_scx_bpf_consume_task(BPF_FOR_EACH_ITER, p))
        break;
```

Reply:
Since that particular kernel task could be consumed by the other CPU, we just consume any task and move on. We could iterate over the loop to find out the exact kernel task but that seems overkill here, and this seems a good enough approximation. What do you think?

Reply:
I see. This seems like a reasonable approximation for now. If there's an issue we can always optimize.